### PR TITLE
Remove elements from cache if they are no longer going to be used

### LINKED ISSUE #920: Perform cache lifetime analysis in Nx.Defn.Evaluator
Today we keep the cache alive forever, but we need to purge entries once they are all used.

### PR DESCRIPTION
**Pull request by josevalim**
Closes #920.

### CODE REVIEW

File: nx/lib/nx/defn/evaluator.ex [hunk 1]
```diff
@@ -42,135 +42,304 @@ defmodule Nx.Defn.Evaluator do
...
-  defp eval(%Nx.Tensor{data: %Expr{op: :tensor, args: [t]}}, _state, cache) do
-    {t, cache}
+  defp precompile(fun, vars, hooks) do
+    expr = fun.(vars)
+    state = %{hooks: hooks, parent_ids: nil, current_ids: nil}
+    cache = init_compute_cache(expr, state, %{})
+    {expr, cache}
+  end
+
+  defp init_compute_cache(expr, state, parent_ids) do
```
Comment (jonatanklosko) [inline nx/lib/nx/defn/evaluator.ex:63]: fwiw currently we always call it with `parent_ids = %{}`.

File: nx/lib/nx/defn/evaluator.ex [hunk 2]
```diff
@@ -42,135 +42,304 @@ defmodule Nx.Defn.Evaluator do
...
+    {elem(tuple, i), caches}
+  end
+
+  defp eval_apply(:attach_token, %Nx.Tensor{data: %Expr{args: [token, expr]}}, state, caches) do
+    {_, caches} = eval(token, state, caches)
+    eval(expr, state, caches)
+  end
+
+  defp eval_apply(:fun, %{data: %Expr{args: [args, expr, _mfa], id: id}}, state, caches) do
+    fun_cache = Map.fetch!(hd(caches), [:fun | id])
```
Comment (jonatanklosko) [inline nx/lib/nx/defn/evaluator.ex:260]: Curious: what's the benefit over `{:fun, id}`?

File: nx/lib/nx/defn/evaluator.ex [hunk 3]
```diff
@@ -42,135 +42,304 @@ defmodule Nx.Defn.Evaluator do
...
 
-  defp eval(%Nx.Tensor{data: %Expr{op: op, id: id}} = ans, state, cache) do
+  defp eval(%Nx.Tensor{data: %Expr{op: op, id: id}} = ans, state, [cache | caches]) do
     case cache do
-      %{^id => res} ->
-        {res, cache}
+      %{^id => count} when is_integer(count) ->
+        {res, [cache | caches]} = eval_apply(op, ans, state, [cache | caches])
+        state.gc && :erlang.garbage_collect(self())
+        {res, [decrement_cache(cache, id, count, res) | caches]}
```
Comment (jonatanklosko) [inline nx/lib/nx/defn/evaluator.ex:189]: Why don't we garbage collect after decrementing (possibly pruning) the cache?

Also, why not garbage collect in the branch below?

File: nx/lib/nx/defn/evaluator.ex (hunk 2 above)
Comment (josevalim) [reply nx/lib/nx/defn/evaluator.ex:260]: It is a (bad?) habit I picked: `[:fun | id]` uses one word fewer compared to tuple, but honestly unlikely to matter here. :)

File: nx/lib/nx/defn/evaluator.ex (hunk 3 above)
Comment (josevalim) [reply nx/lib/nx/defn/evaluator.ex:189]: The value that is being pruned from the cache is `res`, which we are returning from this function, which means it is still alive. In another word, we purge from the catch the last time a value will be used, but it still has to be used.

Comment (jonatanklosko) [reply nx/lib/nx/defn/evaluator.ex:189]: Ah, makes sense!

File: nx/lib/nx/defn/evaluator.ex (hunk 2 above)
Comment (jonatanklosko) [reply nx/lib/nx/defn/evaluator.ex]: Good to know, thanks :) For a moment it made me wonder if `id` isn't a list, it probably sacrifices some readability.

File: nx/test/nx/defn/evaluator_test.exs [hunk 4]
```diff
@@ -478,4 +478,144 @@ defmodule Nx.Defn.EvaluatorTest do
...
+
+    test "with nested map" do
+      assert cond_cache_map(%{iteration: 0}) ==
+               %{iteration: Nx.tensor(1)}
+
+      assert cond_cache_map(%{iteration: 0, abc: 1}) ==
+               %{iteration: Nx.tensor(1), abc: Nx.tensor(1)}
+
+      assert cond_cache_map(%{iteration: 0, xyz: 1}) ==
+               %{iteration: Nx.tensor(1), xyz: Nx.tensor(1)}
```
Comment (jonatanklosko) [inline nx/test/nx/defn/evaluator_test.exs:584]: Aren't these the same?

```suggestion
      assert cond_cache_map(%{iteration: 0, abc: 1}) ==
               %{iteration: Nx.tensor(1), abc: Nx.tensor(1)}
```

File: nx/test/nx/defn/evaluator_test.exs [hunk 5]
```diff
@@ -478,4 +478,144 @@ defmodule Nx.Defn.EvaluatorTest do
...
+
+      assert cond_cache_map(%{iteration: 4}) ==
+               %{iteration: Nx.tensor(5)}
+
+      assert cond_cache_map(%{iteration: 4, abc: 1}) ==
+               %{iteration: Nx.tensor(5), abc: Nx.tensor(1)}
+
+      assert cond_cache_map(%{iteration: 4, xyz: 1}) ==
+               %{iteration: Nx.tensor(5), xyz: Nx.tensor(1)}
```
Comment (jonatanklosko) [inline nx/test/nx/defn/evaluator_test.exs:593]:
```suggestion
      assert cond_cache_map(%{iteration: 4, abc: 1}) ==
               %{iteration: Nx.tensor(5), abc: Nx.tensor(1)}
```

File: nx/test/nx/defn/evaluator_test.exs [hunk 6]
```diff
@@ -478,4 +478,144 @@ defmodule Nx.Defn.EvaluatorTest do
...
+      {state, prev} =
+        if state.iteration < 12 do
+          prev_timestep = if(state.iteration < 2, do: 1, else: -1)
+
+          factor =
+            if prev_timestep >= 0 do
+              2
+            else
+              1
+            end
```
Comment (jonatanklosko) [inline nx/test/nx/defn/evaluator_test.exs:606]:
```suggestion
          sign = if(state.iteration < 2, do: 1, else: -1)

          factor =
            if sign >= 0 do
              2
            else
              1
            end
```

File: nx/test/nx/defn/evaluator_test.exs (hunk 4 above)
Comment (josevalim) [reply nx/test/nx/defn/evaluator_test.exs:584]: No! they changed the map ordering which lead to different errors :D I will drop a comment.

File: nx/test/nx/defn/evaluator_test.exs [hunk 7]
```diff
@@ -478,4 +478,144 @@ defmodule Nx.Defn.EvaluatorTest do
...
+        end
+
+      %{state | iteration: state.iteration + 1}
+    end
+
+    test "with nested map" do
+      assert cond_cache_map(%{iteration: 0}) ==
+               %{iteration: Nx.tensor(1)}
+
+      assert cond_cache_map(%{iteration: 0, abc: 1}) ==
```
Comment (josevalim) [inline nx/test/nx/defn/evaluator_test.exs:580]:
```suggestion
      # Use :abc/:xyz so we try different key orderings.
      assert cond_cache_map(%{iteration: 0, abc: 1}) ==
```

File: nx/test/nx/defn/evaluator_test.exs [hunk 8]
```diff
@@ -478,4 +478,144 @@ defmodule Nx.Defn.EvaluatorTest do
...
+               %{iteration: Nx.tensor(1), abc: Nx.tensor(1)}
+
+      assert cond_cache_map(%{iteration: 0, xyz: 1}) ==
+               %{iteration: Nx.tensor(1), xyz: Nx.tensor(1)}
+
+      assert cond_cache_map(%{iteration: 4}) ==
+               %{iteration: Nx.tensor(5)}
+
+      assert cond_cache_map(%{iteration: 4, abc: 1}) ==
```
Comment (josevalim) [inline nx/test/nx/defn/evaluator_test.exs:589]:
```suggestion
      # Use :abc/:xyz so we try different key orderings.
      assert cond_cache_map(%{iteration: 4, abc: 1}) ==
```

File: nx/test/nx/defn/evaluator_test.exs (hunk 4 above)
Comment (jonatanklosko) [reply nx/test/nx/defn/evaluator_test.exs:584]: Oh wow :smile:

**Review by jonatanklosko — APPROVED** [review]

**Comment by jonatanklosko** [comment]
Beautiful!! :shipit:

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**